### PR TITLE
Add vSphere workers CAPI spec generator

### PR DESCRIPTION
**Pull request by g-gaston**
## Description of changes
Part of the cluster lifecycle project. Generate CAPI spec in yaml with the vsphere template builder and parse it into api structs. Generate new immutable changes if a difference is detected (upgrades).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/clusterapi/controlplane.go [hunk 1]
```diff
@@ -13,7 +13,7 @@ import (
 )
 
 // ControlPlane represents the provider-specific spec for a CAPI control plane using the kubeadm CP provider
-type ControlPlane[C, M Object] struct {
+type ControlPlane[C Object[C], M Object[M]] struct {
```
Comment (jonathanmeier5) [inline pkg/clusterapi/controlplane.go:16]: Checking my understanding here: `Object` itself is now a generic, and we pass `[C]` and `[M]` respectively to specify Object's type?

File: pkg/clusterapi/workers.go [hunk 2]
```diff
@@ -45,7 +47,7 @@ type WorkerGroup[M Object] struct {
 // with the current state of the cluster. If they had, it generates a new name for them by increasing a monotonic number
 // at the end of the name.
 // This process is performed to the provider machine template and the kubeadmconfigtemplate.
-// The kubeadmconfigtemplate is not immutable at the API level but we treat is a such for consistency
+// The kubeadmconfigtemplate is not immutable at the API level but we treat is a such for consistency.
```
Comment (jonathanmeier5) [inline pkg/clusterapi/workers.go:50]:
```suggestion
// The kubeadmconfigtemplate is not immutable at the API level but we treat it as such for consistency.
```

**Comment by jonathanmeier5** [comment]
My understanding of this PR is that the new `WorkersSpec` function uses the old yaml-driven vsphere worker template  builder to generate a new set of CAPI objects based on an EKS-A cluster spec in the management cluster.

After generating the CAPI objects as yaml documents, it parses them into structs and diffs the generated CAPI objects against those in the cluster. If a diff is found, a new name is generated via a monotonically increasing integer appended to the name. The new object is applied, causing an "update".

This all sounds reasonable to me. Naturally I defer to others on the team who have been on the project longer.

/lgtm

File: pkg/clusterapi/controlplane.go (hunk 1 above)
Comment (g-gaston) [reply pkg/clusterapi/controlplane.go:16]: addressed offline

[the remark this answers is not in the document]
Comment (g-gaston) [reply pkg/clusterapi/controlplane.go:16]: Yeah
Now the `Object` interface is typed, so I can define a `DeepCopy` object, that returns the type `C`

**Pull request merged**